\documentclass{bioinfo}

\copyrightyear{2014}

\pubyear{2014}

\begin{document}

\firstpage{1}

\title{Chemically Aware Model Builder (camb): An R package for bioactivity and property modeling of small molecules and proteins}
\author[Murrell \& Cortes-Ciriano \it{et~al}]{Daniel S. Murrell\,$^{1}$\footnote{Equal contributors} , Isidro Cortes-Ciriano\,$^{2,*}$, Gerard J.P. van Westen\,$^{3}$, Andreas Bender$^{1,}$\dag, Th\'er\`ese E. Malliavin\,$^{2,}$\footnote{to whom correspondence should be addressed},  Robert C. Glen$^{1,}$\dag}
\address{$^{1}$Unilever Centre for Molecular Science Informatics, Department of Chemistry, University of Cambridge, Lensfield Road, Cambridge CB2 1EW, United Kingdom.\\
$^{2}$Unite de Bioinformatique Structurale, Institut Pasteur and CNRS UMR 3825, Structural Biology and Chemistry Department, 25\-28, rue Dr. Roux, 75 724 Paris, France.\\
$^{3}$European Molecular Biology Laboratory European Bioinformatics Institute Wellcome Trust Genome Campus, Hinxton, United Kingdom.}

\history{Received on XXXXX; revised on XXXXX; accepted on XXXXX}
\editor{Associate Editor: XXXXXXX}
\maketitle

\begin{abstract}
\section{Summary:}
{\it camb} is an R package that can be used for the rapid generation of quantitative predictive models in the area of medicinal chemistry (QSAR, QSPR, QSAM, proteochemometrics and chemogenomics). It is aimed at both advanced and beginner R users.
Its capabilities include the standardisation and representation of chemical structures, computation of 905 two-dimensional and 14 fingerprint type descriptors for small molecules, the computation of 8 types of amino acid descriptors and 13 different whole protein sequence descriptors, filter methods for feature selection, generation of predictive models (R package {\it caret}), as well as techniques to ensemble these models (R package {\it caretEnsemble}).
Results can be visualised through high-quality, customisable plots (R package {\it ggplot2}).
\section{Availability:} {\it camb} is written in R, C++, Python and Java and is available open source
at https://github.com/cambDI/camb.
Two tutorials are also included.\\
\section{Contact:} Therese Malliavin: terez@pasteur.fr or Robert Glen: rcg28@cam.ac.uk
% [TBD: link to the github repo at some place in this document]
\end{abstract}

\section{Introduction}

The advent of high-throughtput technologies over the last two decades 
has led to a vast increase of compound, bioactivity
and genomic databases \citep{bender_databases} [TBD: cite other databases].
This rampant increase in the amount of chemical and biological information 
has been exploited by emergent fields in drug discovery 
such as quantitative structure activity relationships (QSAR), quantitative structure property relationships (QSPR), quantitative sequence-activity modelling (QSAM), or proteochemometric modelling (PCM) \citep{review_pcm}.

The R programming language provides a flexible platform for statistical analyses \citep{Rlanguage},
and its applicability in medicinal chemistry has been reviewed elsewhere \citep{mente}.
Although R is extensively used in diverse biological domains, {\it e.g.} genomics \citep{bioconductor},
the availability of R packages for cheminformatics and medicinal chemistry is limited. % \citep{mente}.
Nonetheless, R still constitutes the most frequent choice in the medicinal chemistry literature
for compounds bioactivity and property modelling \citep{mente}.
In general, these studies share a common algorithmic structure, which can be summarised in 4 model generation steps:
(i) compound standardisation, (ii) descriptor calculation,
(iii) preprocessing, feature selection, model training and validation, and (iv) bioactivity/property prediction for new molecules.

Currently available R packages provide the capability for a subset of the previous steps.
For instance, R packages {\it chemmineR} \citep{chemmineR} and {\it rcdk} \citep{rcdk} enable the manipulation of SDF and SMILES
files, the calculation of physicochemical descriptors, the clustering of molecules,
and the retrieval of compounds from PubChem \citep{pubchem}.
On the machine learning side, the {\it caret} package provides a
unified platform for the training of machine learning models \citep{caret}.

Here, we present the R package {\it camb}: {\bf C}hemically {\bf A}ware {\bf M}odel {\bf B}uilder,
which aims to address the current lack of an R framework encompasing all four steps mentioned above.
The package has been conceived in a way that users with minimal
programming skill are able to generate competitive predictive models and high-quality plots
under default operation.
However, each function can be utilised to fulfil the more versatile needs of more experienced users.

Overall, {\it camb} enables the generation of predictive  models (QSAR, QSPR, QSAM and PCM)
starting with: chemical structure files, protein sequences (if required), and the associated properties or bioactivities.
Moreover, {\it camb} is the first R package that enables the manipulation of chemical structures utilising Indigo's C API \citep{Indigo},
and the calculation of:
(i) molecular fingerprints and descriptors calculated using the PaDEL-Descriptor Java library (v2.16) \citep{padel},
(ii) hashed and unhashed Morgan fingerprints \citep{extended_fp},
and (iii) 8 types of amino acid descriptors [TBD: are these implemented here for the first time?]. 
Two case studies illustrating the application of {\it camb} for
QSPR modelling (solubility prediction) and PCM are available in the online supplementary information.
In the following section we detail the main functionalities provided by {\it camb}. 

\section{Description}
This section describes the tools provided by {\it camb} 
for (i) compound standardisation, (ii) molecular descriptor calculation, 
(iii) preprocessing and feature selection, model training, visualisation and validation, and (iv) bioactivity/property prediction for new molecules.	

\subsection{Compound stardardization}

In order to represent all molecules in a given dataset in the same 
way (compound standardisation),
{\it camb}  provides the function {\it StandardiseMolecules} which utilises Indigo's C API \citep{Indigo}.
SDF and SMILES formats are provided as molecule input options.
The maximum number of fluorines, chlorines, bromines and iodines
that a compound can include in order to pass the standardisation process can be defined by the user.
Additional arguments of this function include the removal of inorganic molecules
or those compounds with a molecular mass above or below a given thresholds.
{\it camb} makes use of Indigo's InChI [TBD: cite] InChI plugin to standardise tautomers to the same SMILES representation
by converting to InChI, discarding tautomeric information, and converting back to SMILES.

\subsection{Descriptor calculation} 

Currently, {\it camb} supports the calculation of compound descriptors and fingeprints via PaDEL-Descriptor \citep{padel},
and extended connectivity circular fingerprints \citep{extended_fp} as implemented in the RDkit \citep{rdkit}.
The function {\it GeneratePadelDescriptors} permits the calculation of 905 2-dimensional descriptors and 10 PaDEL fingerprints [TBD: mention fingerprints].
%and the following fingerprints:
%CDK fingerprint, CDK extended fingerprint, Estate fingerprint \cite{state_fp}, CDK graph only fingerprint, MACCS fingerprint,
%Pubchem fingerprint, Substructure fingerprint, Substructure fingerprint count \citep{obabel}, Klekota-Roth fingerprint and Klekota-Roth fingerprint count \citep{privileged_substructures}.

Morgan fingerprints can be computed with the function {\it MorganFPs}
through the python library RDkit \citep{rdkit}.
Hashed fingerprints are either binary, recording the presence or absence of each substructure,
 or count based, recording the number of occurrences of each substructure.
Additionally, this function computes unhashed (keyed) fingerprints, 
where each substructure in the dataset is assigned a unique position in a binary fingerprint of length equal to the number of substructures existing in the dataset.

Since the positions of substructures in the unhashed fingerprint depend on the dataset, the function {\it MorganFPs} allows calculation of unhashed fingerprints for new compounds using a basis defined by the substructures present in the training dataset. This ensures that substructures in new compounds map to the same locations on the fingerprint.

{\it camb} also enables the calculation of 13 types of whole protein sequence descriptors
from UniProt identifiers \citep{protr},
as well as the calculation of 8 types of amino acid descriptors \citep{AA_benchmark}.

\subsection{Model training and validation}

Prior to model training, descriptors often need to be statistically preprocessed \citep{andersson}.
To this end, several functions (see package documentation and tutorials)
are provided, {\it e.g.} the removal of non-informative predictors or
their conversion to z-scores.

{\it camb} invokes the R package {\it caret} to set up cross validation frameworks and
train individual machine learning models. {\it caret} provides a common interface to the most popular machine learning package that exist in R.
These include learning methods in Bagging, Bayesian Methods, Boosting, Boosted Trees, Elastic Net, MARS, Gaussian Processes, K Nearest Neighbour,
 Principal Component Regression, Radial Basis Function Networks, Random Forests, Relevance Vector Machines, Support Vector Machines among others.
Additionally, two ensemble modelling approaches, namely: greedy and stacking optimisation,
have been integrated from the R package {\it caretEnsemble} \citep{caretEnsemble}, which allow a combination of models to be used 
allowing for more accurate models to be built.
Statistical metrics for model validation have also been included \citep{beware}.

Some visualization functionality is provided. All plots are generated using the R package {\it ggplot2} \citep{ggplot2}.
Default options for plotting functions allow the generation of high-quality plots,
however, the layer-based structure of ggplot objects allows for further optimisation
by the addition of customisation layers.  
Visual depiction of compounds is also possible with the function {\it PlotMolecules},
utilising Indigo's C API. Visualization functions are exemplified in the tutorials provided in the supplementary information.

\subsection{Predictions for new molecules}
One of the major benefits of having all the tools available in one framework is that it makes it easy to run new molecules through exactly the same routines that the training set of molecules was subjected to before the model training process. {\it camb} provides the option to provide a test set of molecules along with a trained model, and outputs predictions on the test set.

\section{Conclusions}
{\it In silico} predictive models have proved valuable
for the optimisation of compound potency, selectivity and safety profiles.
In this context, {\it camb} provides a complete framework
to (i) manipulate compound structures, (ii) generate compound and protein descriptors,
(iii) train and validate 
QSAR, QSPR, QSAM, PCM and chemogenomic models, and
(iv) make predictions for new molecules.

\section{Acknowledgements}
ICC thanks the Paris-Pasteur International PhD Programme for funding.
ICC and TM thank CNRS, Institut Pasteur and ANR bipbip for funding.
DSM and RCG thanks Unilever for funding.
GvW thanks EMBL (EIPOD) and Marie Curie (COFUND) for funding.
AB thanks Unilever and the European Research Commission (Starting Grant ERC-2013-StG 336159 MIXTURE) for funding.

\bibliographystyle{natbib}
\bibliography{biblio}

\end{document}

% [TBD: try and get recursive installation of package working when this one is installed]
%\begin{figure}[htb!]
%\begin{center}
%\includegraphics[width=\textwidth]{camb_outline.pdf}
%\end{center}
%\caption{}
%\label{scheme}
%\end{figure}
