\documentclass{bioinfo}

\copyrightyear{2014}

\pubyear{2014}

\begin{document}

\firstpage{1}

\title{Chemistry Aware Model Builder (camb): An R package for bioactivity and property modeling of small molecules and proteins}
\author[Murrell \& Cortes-Ciriano \it{et~al}]{Daniel S. Murrell\,$^{1}$\footnote{Equal contributors} , Isidro Cortes-Ciriano\,$^{2,*}$, Gerard J.P. van Westen\,$^{3}$, Th\'er\`ese E. Malliavin\,$^{2,}$\footnote{to whom correspondence should be addressed} , Andreas Bender$^{1,}$\dag}
\address{$^{1}$Unilever Centre for Molecular Science Informatics, Department of Chemistry, University of Cambridge, Cambridge, United Kingdom.\\
$^{2}$Unite de Bioinformatique Structurale, Institut Pasteur and CNRS UMR 3825, Structural Biology and Chemistry Department, 25\-28, rue Dr. Roux, 75 724 Paris, France.\\
$^{3}$European Molecular Biology Laboratory European Bioinformatics Institute Wellcome Trust Genome Campus, Hinxton, United Kingdom.}

\history{Received on XXXXX; revised on XXXXX; accepted on XXXXX}
\editor{Associate Editor: XXXXXXX}
\maketitle

\begin{abstract}
\section{Summary:}
{\it camb} is an R package that can be used for the rapid generation of quantitative predictive models in the area of medicinal chemistry (QSAR, QSPR, QSAM, proteochemometrics and chemogenomics). It is aimed at both amateur and advanced R users.
Its capabilities include the standardisation and representation of chemical structures, computation of 905 two-dimensional and 14 fingerprint type descriptors for small molecules, the computation of 8 types of amino acid descriptors and 13 different whole protein sequence descriptors, filter based statistical preprocessing, generation of predictive models (R package {\it caret}), as well as techniques to ensemble these models (R package {\it caretEnsemble}).
Results can be visualised through high-quality, customisable plots (R package {\it ggplot2}).
\section{Availability:} {\it camb} is written in R, C++, Python and Java and is available open source
at https://github.com/cambDI/camb.
Two tutorials are also included.\\
\section{Contact:} dsmurrell@gmail.com or isidrolauscher@gmail.com
\end{abstract}

\section{Introduction}

The advent of high-throughtput technologies over the last two decades 
has led to a vast increase of compound, bioactivity
and genomic databases \citep{bender_databases}.
This rampant increase in the amount of chemical and biological information 
has been exploited by emergent fields in drug discovery 
such as chemogenomics or proteochemometrics (PCM) \citep{review_pcm,cortesReview}.

The R programming language provides a excellent platform for statistical analyses \citep{Rlanguage},
and its applicability in medicinal chemistry has been reviewed elsewhere \citep{mente}.
Although R is extensively used in diverse biological domains, {\it e.g.} genomics \citep{bioconductor},
the availability of R packages for cheminformatics and medicinal chemistry is limited. % \citep{mente}.
Nonetheless, R still constitutes the most frequent choice in the medicinal chemistry literature
for compounds bioactivity and property modelling \citep{mente}.
In general, these type of studies share a common structure, which can be summarised in 4 model generation steps:
(i) compound standardisation, (ii) descriptor calculation,
(iii) preprocessing, model training and validation, and (iv) bioactivity/property prediction for new molecules.

Currently available R packages provide the capability for a subset of the previous steps.
For instance, R packages {\it chemmineR} \citep{chemmineR} and {\it rcdk} \citep{rcdk} enable the manipulation of SDF and SMILES
files, the calculation of physicochemical descriptors, the clustering of molecules,
and the retrieval of compounds from PubChem \citep{pubchem}.
On the machine learning side, the {\it caret} package provides a
unified platform for the training of machine learning models \citep{caret}.

Here, we present the R package {\it camb}: {\bf C}hemistry {\bf A}ware {\bf M}odel {\bf B}uilder,
which aims to address the current lack of an R framework encompasing all four steps mentioned above.
The package has been conceived in a way that users with little
programming skill are able to generate competitive predictive models and high-quality plots
under default operation.
However, each function can be utilised to fulfil the more versatile needs of more experienced users.

Overall, {\it camb} enables the generation of predictive  models (QSAR, QSPR, QSAM, PCM and chemogenomics)
starting from chemical structure files, optional protein sequences, and the associated properties or bioactivities.
Moreover, {\it camb} is the first R package enabling the manipulation of chemical structures {\it via} the C-written Indigo API \citep{Indigo},
and the calculation of:
(i) PaDEL descriptors and fingerpints \citep{padel},
(ii) hashed and unhashed Morgan fingerprints \citep{extended_fp},
and (iii) 8 types of amino acid descriptors. 
Two case studies illustrating the application of {\it camb} for
QSPR modelling and PCM are available in the online supplementary information.
In the following section we detail the main functionalities provided by {\it camb}. 

\section{Description}
This section describes the tools provided by {\it camb} 
for (i) compound standardisation, (ii) descriptor calculation, 
(iii) preprocessing, model training and validation, and (iv) visualisation. % these differ from the 4 steps highlighted above	

\subsection{Compound stardardization}

In order to represent all molecules in a given dataset in the same 
way (compound standardisation),
{\it camb}  provides the function {\it StandardiseMolecules} which utilises Indigo's C API \citep{Indigo}.
SDF and SMILES formats are provided as molecule input options.
The maximum number of fluorines, chlorines, bromines and iodines
that a compound can exhibit in order to pass the standardisation process can be defined by the user.
Additional arguments of this function include the removal of inorganic molecules
or those compounds with a molecular mass above or below a given thresholds.
{\it camb} makes use of Indigo's InChI plugin to standardise tautomers to the same SMILES representation
by converting to InChI, discarding tautomeric information, and converting back to SMILES.

\subsection{Descriptor calculation} 

Currently, {\it camb} supports the calculation of compounds descriptors and fingeprints from PaDEL \citep{padel},
and circular Morgan fingerprints \citep{extended_fp} as implemented in the RDkit \citep{rdkit}.
The function {\it GeneratePadelDescriptors} permits the calculation of 905 2-dimensional descriptors and 10 PaDEL fingerprints.
%and the following fingerprints:
%CDK fingerprint, CDK extended fingerprint, Estate fingerprint \cite{state_fp}, CDK graph only fingerprint, MACCS fingerprint,
%Pubchem fingerprint, Substructure fingerprint, Substructure fingerprint count \citep{obabel}, Klekota-Roth fingerprint and Klekota-Roth fingerprint count \citep{privileged_substructures}.

Morgan fingerprints can be computed with the function {\it MorganFPs}
through the python library RDkit \citep{rdkit}.
Hashed fingerprints are calculated in binary format and with counts.
Additionally, this function computes unhashed (keyed) fingerprints.
In this case, each substructure in the dataset is assigned a unique position in a binary fingerprint.
To calculate the fingerprint for each compound those positions in the fingerprint corresponding to the substructures 
present in a given compound are set to 1 (binary format)
or the number of times the substructure appears in that compound (counts format).

From the above, it is apparent that the position of the set bits in an unhashed fingerprint
directly depends on the dataset.
To facilitate the application of predictive models trained on unhashed fingerprints,
the function {\it MorganFPs} also allows the calculation of unhashed fingerprints for new compounds
using a basis defined by the substructures present in a given chemical training set.

{\it camb} also enables the calculation of 13 types of whole protein sequence descriptors
from UniProt identifiers \citep{protr},
as well as the calculation of 8 types of amino acid descriptors \citep{AA_benchmark}.

\subsection{Model training and validation}

Prior to model training, descriptors should to be statistically preprocessed \citep{andersson}.
To this aim, several functions (see package documentation and tutorials)
are provided, {\it e.g.} the removal of non-informative predictors or
their conversion to z-scores.

{\it camb} invokes the R package {\it caret} to 
train individual machine learning models.
Additionally, two ensemble modelling approaches, namely: greedy and stacking optimisation,
have been integrated from the R package {\it caretEnsemble} \citep{caretEnsemble}.
Statistical metrics for model validation have also been included \citep{beware}.

\subsection{Visualization}
All plots are generated using the R package {\it ggplot2} \citep{ggplot2}.
Default options for plotting functions allow the generation of 
high-quality plots,
however, the layer-based structure of ggplot objects allows for further tweaking
by the addition of customisation layers.  
Visual depiction of compounds is also possible with the function {\it PlotMolecules},
utilises Indigo's C API.
Visualization functions are explained in the tutorials.

\section{Conclusions}
{\it In silico} predictive models have proved a valuable
tool for the optimisation of compounds potency, selectivity and safety profiles.
In this context, {\it camb} provides a complete framework
to (i) manipulate compound structures, (ii) generate compound and protein descriptors, and
(iii) train and validate 
QSAR, QSPR, QSAM, PCM and chemogenomic models.

\section{Acknowledgements}
ICC thanks the Paris-Pasteur International PhD Programme for funding.
ICC and TM thank CNRS, Institut Pasteur and ANR bipbip for funding.
DM thanks Unilever for funding.
GvW thanks EMBL (EIPOD) and Marie Curie (COFUND) for funding.
AB thanks Unilever and the European Research Commission (Starting Grant ERC-2013-StG 336159 MIXTURE) for funding.

\bibliographystyle{natbib}
\bibliography{biblio}

\end{document}

%\begin{figure}[htb!]
%\begin{center}
%\includegraphics[width=\textwidth]{camb_outline.pdf}
%\end{center}
%\caption{}
%\label{scheme}
%\end{figure}
